torch/distributed/distributed_c10d.py

```diff
@@ -436,6 +436,10 @@ def init_process_group(backend,
     _backend = _pg_map[_default_pg][0]
     _default_pg_init_method = init_method
 
+    # barrier at the end to ensure that once we return from this method, all
+    # process groups including global variables are updated correctly on all
+    # ranks.
+    barrier()
```

Comment:
cc @agolynski FYI, after this change, c10d extensions cannot throw in `barrier()`, otherwise they won't be able to initialize default process groups.